protocol/jsm/v2/receiver.go

```diff
@@ -0,0 +1,164 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package jsm
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+type msgErr struct {
+	msg binding.Message
+}
+
+type Receiver struct {
+	incoming chan msgErr
+}
+
+func NewReceiver() *Receiver {
+	return &Receiver{
+		incoming: make(chan msgErr),
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (r *Receiver) MsgHandler(msg *nats.Msg) {
+	r.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+func (r *Receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+type Consumer struct {
+	Receiver
+
+	Conn       *nats.Conn
+	Jsm        nats.JetStreamContext
+	Subject    string
+	Subscriber Subscriber
+
+	subMtx        sync.Mutex
+	internalClose chan struct{}
+	connOwned     bool
+}
+
+func NewConsumer(url, stream, subject string, natsOpts []nats.Option, jsmOpts []nats.JSOpt, opts ...ConsumerOption) (*Consumer, error) {
+	conn, err := nats.Connect(url, natsOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := NewConsumerFromConn(conn, stream, subject, jsmOpts, opts...)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	c.connOwned = true
+
+	return c, err
+}
+
+func NewConsumerFromConn(conn *nats.Conn, stream, subject string, jsmOpts []nats.JSOpt, opts ...ConsumerOption) (*Consumer, error) {
+	jsm, err := conn.JetStream(jsmOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	streamInfo, err := jsm.StreamInfo(stream, jsmOpts...)
+
+	if streamInfo == nil && err != nil {
+		_, err = jsm.AddStream(&nats.StreamConfig{
```

Comment:
maybe here it would help to ignore the error in case it exists so that it would just create the consumer? Something like:
```go
if err != nil && !strings.Contains(err.Error(), "already in use") {
```

Reply:
Sounds like a better way